crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -438,4 +453,144 @@ struct EthApiInner<Provider, Pool, Network> {
     pending_block: Mutex<Option<PendingBlock>>,
     /// A pool dedicated to blocking tasks.
     blocking_task_pool: BlockingTaskPool,
+    /// Cache for block fees history
+    fee_history_cache: FeeHistoryCache<Provider>,
+}
+
+/// Settings for the [EthStateCache](crate::eth::cache::EthStateCache).
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FeeHistoryCacheConfig {
+    /// Max number of blocks in cache.
+    ///
+    /// Default is 1024.
+    pub max_blocks: u64,
+}
+
+impl Default for FeeHistoryCacheConfig {
+    fn default() -> Self {
+        FeeHistoryCacheConfig { max_blocks: 1024 }
+    }
+}
+
+/// Wrapper struct for BTreeMap
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache<Provider> {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    provider: Provider,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, SealedHeader>>>,
```

Comment:
this should not store the entire sealedheaders but the only what we need, basefee, gas_used ratio, and then the percentiles